buildSrc/subprojects/plugins/src/main/kotlin/org/gradle/plugins/performance/PerformanceTestPlugin.kt

```diff
@@ -417,37 +403,35 @@ class PerformanceTestPlugin : Plugin<Project> {
             maxParallelForks = 1
 
             project.findProperty(PropertyNames.baselines)?.let { baselines ->
-                systemProperty(PropertyNames.baselines, baselines)
+                task.baselines = baselines as String
             }
 
             jvmArgs("-Xmx3g", "-XX:+HeapDumpOnOutOfMemoryError")
 
             dependsOn(prepareSamplesTask)
-            finalizedBy(performanceReportTask)
+
+            registerTemplateInputsToPerformanceTest()
 
             mustRunAfter(tasks.withType<ProjectGeneratorTask>())
             mustRunAfter(tasks.withType<RemoteProject>())
             mustRunAfter(tasks.withType<JavaExecProjectGeneratorTask>())
-
-            doFirst {
-                channel?.let { channel ->
-                    performanceReportTask.get().systemProperty(PropertyNames.channel, channel)
-                }
-            }
         }
     }
 
     private
-    fun Project.configureForAnyDistributedPerformanceTestTask(task: DistributedPerformanceTest) {
-        task.apply {
-            val registerInputs: (Task) -> Unit = { prepareSampleTask ->
-                val prepareSampleTaskInputs = prepareSampleTask.inputs.properties.mapKeys { entry -> "${prepareSampleTask.name}_${entry.key}" }
-                task.inputs.properties(prepareSampleTaskInputs)
-            }
-            tasks.withType<ProjectGeneratorTask>().forEach(registerInputs)
-            tasks.withType<RemoteProject>().forEach(registerInputs)
-            tasks.withType<JavaExecProjectGeneratorTask>().forEach(registerInputs)
+    fun DistributedPerformanceTest.configureReportProperties() {
+        reportDir = project.buildDir / Config.performanceTestReportsDir
+    }
+
+    private
+    fun PerformanceTest.registerTemplateInputsToPerformanceTest() {
+        val registerInputs: (Task) -> Unit = { prepareSampleTask ->
+            val prepareSampleTaskInputs = prepareSampleTask.inputs.properties.mapKeys { entry -> "${prepareSampleTask.name}_${entry.key}" }
+            inputs.properties(prepareSampleTaskInputs)
```

Comment:
I don't think we should be doing this (adding the properties from one task as properties of another task via the runtime API).  

This is causing breakages with the 5.0-nightly because we're validating runtime API properties now:
https://e.grdev.net/s/pflxxjtcgnb2k/failure?openFailures=WzBd&openStackTraces=WzEse31d#top=0
https://e.grdev.net/s/gpjl3hssm4dcq/console-log#L1572

This is bad in a few ways:
- There's a legitimate problem with `BuildBuilderGenerator` mixing Integer/Strings
- There failures are nonsensical.  You're seeing task property validation for another task's properties.
- The properties we're taking from the sample task can sometimes be optional.  We don't have a way to express that here, so we get failures.

Shouldn't the inputs to the PerformanceTest just be the sample directory (as files)?

Reply:
Thanks @big-guy . We don't like this, either. The reason why we can't register sample directories as `PerformanceTest`'s input is, `PerfomanceTestCoordinator` doesn't generate any sample projects at all, it just schedules workers to do the real work via TC REST API. We probably have two choices:

- Manually register all properties of other tasks by invoking their getters. The cons are that when people add properties into other tasks, they might forgot to update `PerformanceTest`.

- Do not cache `PerformanceTestCoordinator`. Individual performance test workers can register sample project as input. This requires sample project generator to be deterministic.

Reply:
For the individual performance test runners it is not completely easy to register a template as an input, given that not all templates are generated for an individual performance test, but only the template required by that test. We currently don't have a way to know that up front. It would be good though, maybe we should introduce another task which determines the necessary templates from the selected scenarios to run.

For the time being, I think that every template should implement an interface which allows to add its inputs to the performance test job. I think that should be good enough for now.

Another option would be to generate all the templates for the coordinator and use those as an input. If we make the generation cacheable for certain kinds of templates, that could be feasible, too.

The individual tests then could have the templates generated as part of the build as an input.

Reply:
Here is a PR for allowing null inputs from the template tasks: https://github.com/gradle/gradle/pull/6599

And here is another one adding the actual generated projects as an input to the performance test: https://github.com/gradle/gradle/pull/6600

We can unblock the `master` wrapper upgrade by merging #6599.

The overhead of #6600 is that we generate all templates for the distributed performance tests. That takes about 6 minutes on our CI infrastructure: https://e.grdev.net/s/5wvl2bwxbznpc/timeline
We can lower that time by improving cacheability of the project generator tasks.

Reply:
Thanks @wolfs , I think #6599 is what we need right now, but #6600 should be discussed a little more.

I don't know why we have been generating templates since long time ago. In my understanding, templates can be pre-generated and stored in zip or uploaded to Github as a repository, and we just need to download/unzip/clone it when necessary. This at least has a few advantages:

- If we are not using it (as in this performance coordinator case), we don't need to spend a lot of time/disk IO on these template. We can only check its revision or `HEAD` commit.
- Reduce the overhead of generator task.
- Can remove the generator tasks out of our codebase, or at least remove them out of `buildSrc`.

Reply:
> templates can be pre-generated and stored in zip 

That's what we call the build cache :)

Overall I think we should move the generator code out of buildSrc and into our main code base, so that the performance tests can just generate the project they want to run themselves. This would improve their usability. We've already partially done this with the Java project generators (e.g. largeJavaMultiProject).

As a result, any change to the templates would simply be registered as a classpath change and trigger a rerun of the performance test.